testsuite/tests/eventring/stubs.c

```diff
@@ -0,0 +1,105 @@
+#define CAML_NAME_SPACE
+
+#include "caml/alloc.h"
+#include "caml/eventring.h"
+#include "caml/fail.h"
+#include "caml/memory.h"
+#include "caml/mlvalues.h"
+
+#include <assert.h>
+
+struct counters {
+    int minor_started;
+    int major_started;
+    int compact_started;
+    int minors;
+    int majors;
+    int compacts;
+};
+
+void start_eventring() {
+    caml_eventring_start();
+}
+
+int ev_begin(int domain_id, void* callback_data,
+                uint64_t timestamp, ev_runtime_phase phase) {
+    struct counters* tmp_counters = (struct counters*)callback_data;
+    switch( phase ) {
+        case EV_MINOR:
+            tmp_counters->minor_started = 1;
+            break;
+        case EV_MAJOR:
+            tmp_counters->major_started = 1;
+            break;
+        case EV_COMPACT_MAIN:
+            tmp_counters->compact_started = 1;
+            break;
+        default:
+            break;
+    }
+
+    return 1;
+}
+
+int ev_end(int domain_id, void* callback_data, uint64_t timestamp,
+                ev_runtime_phase phase) {
+    struct counters* tmp_counters = (struct counters*)callback_data;
+    switch( phase ) {
+        case EV_MINOR:
+            assert(tmp_counters->minor_started);
+            tmp_counters->minor_started = 0;
+            tmp_counters->minors++;
+            break;
+        case EV_MAJOR:
+            assert(tmp_counters->major_started);
+            tmp_counters->major_started = 0;
+            tmp_counters->majors++;
+            break;
+        case EV_COMPACT_MAIN:
+            assert(tmp_counters->compact_started);
+            tmp_counters->compact_started = 0;
+            tmp_counters->compacts++;
+            break;
+        default:
+            break;
+    }
+
+    return 1;
+}
+
+value get_event_counts(void) {
+    CAMLparam0();
+    CAMLlocal1(counts_tuple);
+    eventring_error res;
+    uintnat events_consumed;
+
+    struct counters tmp_counters = { 0 };
+
+    counts_tuple = caml_alloc_small(3, 0);
+
+    struct caml_eventring_cursor* cursor;
+
+    res = caml_eventring_create_cursor(NULL, -1, &cursor);
+
+    if( res != E_SUCCESS ) {
+        caml_failwith("invalid or non-existent cursor");
```

Comment:
```suggestion
        caml_failwith("Eventring.get_event_counts: invalid or non-existent cursor");
```

It's worth prefacing the caml_failwiths in this file with `Eventring: ` so that the exception is a little more descriptive.

Reply:
Yes, good idea. Incoming commit that does that.

Is there a convention for the exceptions `eventring.c` and `eventring_consumer.c` raise? Some parts of the runtime seem to be function prefixed, others are file or module.

Reply:
All exceptions from Eventring should have something mentioning eventring prefixed now.